fix issue with infinite redirects

gyzerok opened the pull request:
Hello @jariz!

First of all - amazing plugin! Was really happy to find it out especially since I need proper redirects for SEO. Thanks for your work on it 😃 

**The problem**

It seems that the current implementation for the routing rules for redirects done with `createRedirect` result in infinite redirect loops. Consider following example

```js
createPage({
  path: '/list/1'
  // ...other parameters
})

createRedirect({
  fromPath: '/list',
  toPath: '/list/1',
  isPermanent: true,
  redirectInBrowser: true
})
```
Now if we use `curl` to see how redirects work we will see the following
```bash
$ curl -I https://example.com/list
HTTP/2 301
content-length: 0
location: https://example.com/list/1/

$ curl -I https://example.com/list/1/
HTTP/2 301
content-length: 0
location: https://example.com/list/1/1/

....
```
If you try to open the https://example.com/list url directly in the browser (not from the link on the opened frontend) then you'll get an error and `https://example.com/list/1/1/1/1/1/1/1/1/1/1/1/...` in your browser address bar.

It seems that this is happening because we don't put `HttpErrorCodeReturnedEquals: '404'` for redirects.

**The solution**

For our porject I've needed to do a hotfix while waiting for your response. That's why I am creating a PR here and not an issue.

Basically my change make it so that for exact redirects we will use `HttpErrorCodeReturnedEquals: '404'` and `ReplaceKeyWith`. For redirects with `*` (not sure if these are allowed in Gatsby) we will not use `HttpErrorCodeReturnedEquals: '404'` and use `ReplaceKeyPrefixWith`.

I am no expert in AWS nor in Gatsby. So I might have misunderstood the documentations terribly wrong and it is actually my mistake not a bug in the plugin. If so can you, please, guide me to the right solution?

jariz commented:
The original code was basically the result of nothing but trial and error, the AWS docs aren't very helpful in that regard. (routing rules are ridiculously complicated - for no real reason if you ask me)
I'll test this out somewhere this week to make sure it doesn't break anything else.

gyzerok commented:
Yeah, had the same experience today trying to figure out what's wrong. I am now using my fork for the project and it seems to work as expected at least there. Looking forward for your test results.

I've also changed `withTrailingSlash` function which was not part of the problem itself, but resulted in the generation of double trailing slashes in some cases.

gyzerok commented:
Sorry, I've happened to messed up a bit in the PR. Due to the fact that you use `prepublish` hook to compile things I had to create couple of branches to be able to install from my fork. And it resulted in some changes being out of sync. Fixed now. Hopefully you haven't run your tests yet :)

jariz commented:
@gyzerok 

So, a couple problems I found by deploying the `examples/with-redirects` project from this branch:

- rewrites are now infinite redirecting instead 😅
```
$ curl -I -L http://gatsby-plugin-s3-dev.s3-website-eu-west-1.amazonaws.com/client-only/test

HTTP/1.1 302 Moved Temporarily
Location: http://gatsby-plugin-s3-dev.s3-website-eu-west-1.amazonaws.com/client-only/test

HTTP/1.1 302 Moved Temporarily
Location: http://gatsby-plugin-s3-dev.s3-website-eu-west-1.amazonaws.com/client-only/test

...etc...
```
- The root redirect does not appear to work anymore, `index.html` does not get the WebsiteRedirectLocation metadata key applied to it anymore for some reason.
https://github.com/jariz/gatsby-plugin-s3/blob/master/src/gatsby-node.ts#L57-L77

jariz commented:
Would be really nice if you could fix these 2 problems because I just ran into a similar infinite redirect issue as you were having, and this branch does seem to fix that problem, so I'd really like to get this merged.

gyzerok commented:
@jariz 

To be honest I was not sure what rewrites are at all. And was unable to find explanation in the Gatsby docs. So if you can point me to an explanation it will help me to understand better what the hell am I doing.

It seems that we need 404 code in rewrites as well. You had it before and I added it now as well in the last commit.

Currently my problem is - I do not have any opportunity to test this. I am not familiar with AWS and previously there was a person next to me helping with original PR. Do you have some bucket you are using for tests? Can you create some credentials for me with which I can try deployments and test my PR? Or maybe we could just write some CI script which would deploy examples and then check with curl the redirects?

gyzerok commented:
Of course you can check branch with my last commit now and if everything is good merge it. However it'd be nice anyway to have some automated tests in the repo :)

jariz commented:
@gyzerok Nope, still the same problems.

Client rewrites are client only routes, as the name might suggest, they only exist in the client application.
The do not correspond to a file on the serverside, so if you refresh, you would get a 404 without any redirects in place.
Therefore, any request that is prefixed with the client only route, should redirect to the index of the client only route.

Anyway, I'll take a look at it myself this week.
Might even fix #30 along with it.

gyzerok commented:
@jariz how much time would it take for you to set up the CI for you? I think for the testing and redirect checks we could use just bash with curl. It seems that I might also need the rewrites to work very soon. And I am happy to try to finish up my PR, just need a way of testing it.

jariz commented:
Hi @gyzerok, I've committed the tests to your branch.
I'm still figuring out how to do things on an actual CI (security is kind of a problem that I still need to figure out, because anyone can theoretically make a pull request and have full S3 access to my personal account 🙃)
So for now, you'll have to run it manually on your local machine instead with a valid AWS access token configured.

Ensure you have `npm run develop` running, and repeat the following everytime you want to run a test:

from `cd examples/with-redirects`
```
npm run build && npm run deploy
```

from the project root:
```
TESTING_ENDPOINT=http://replace-me-please.s3-website-eu-west-1.amazonaws.com/ npm t
```

Don't forget to change the bucketname to something more unique in the `with-redirects` project and update the TESTING_ENDPOINT var along with it.

gyzerok commented:
@jariz 

Why root redirect isn't working is a mystery to me. I literally changed nothing in my PR which affects that logic. However I was able to fix rewrites.

Even if I comment out all the redirects the root one doesn't work. I suspect that something within the dependencies have changed, but not sure.

One thing I discovered along the way is that doing rewrites with redirects might be a wrong thing. Consider following example:

Let's imagine in our gatsby project we have user pages `/user/:id`. We can't generate them statically, because we fetch user at the runtime. And we don't know which ids exist in the compile time. If I understand correctly this is exactly the use case for rewrites.

So what we do instead - we have client only paths `/user/*`, which we are planning to resolve using client routing. In this case what we actually want is to return html of `/user` page for any request to `/user/:id` with status 200. However if we redirect from `/user/:id` to `/user` we gotta loose this `:id` url parameter and won't be able to fetch the user we want. On the contrary if we straight away return html, then client routing will kick in and render proper route with param in it.

In your example with `/client-only/secret` when you go straight to the url you gonna loose `/secret` in it. This is how it was working before and I fixed it to work the same way. However I think that it's not an expected behavior.

Thoughts?

jariz commented:
The point of client rewrites is to ensure that when people refresh from a client only route, they are not greeted with a 404. Yes, everything after the client-only prefix will be lost, but that's better than a 404.
You make a good point about returning HTML instead, and if that was an option, I'd say go for it, but with S3 website hosting alone that really can't be accomplished as far as I know.

A better example of client only routes would be a `/my-account/` page, with options like a shopping cart and a user settings page: each of these differ for every user, so there's no point of them existing at all on the server side. `/my-account/` will just show a (statically rendered) login page, and everything prefixed by it requires a login to function.

Regarding the index redirect, maybe it got unknowingly broken in an earlier commit before your work.
Good thing we have tests now I guess 😛
I'll take a look at it.

gyzerok commented:
Yeah, I was expecting this reasoning from you about client rewrites, but wasn't sure. That's why decided to check explicitly. Anyway I fixed this one to work.

Also, please notice that I fixed tests a little bit. Now `TESTING_ENDPOINT` should not include the trailing slash. I've thought that makes a bit more sense.

I will leave the root url for you to figure out then :)

gyzerok commented:
Do you think btw we can merge the PR currently? Or is root redirect something that you use yourself? I am moving to release in my project and would love to switch back to the npm distribution.

jariz commented:
all green 😄

gyzerok commented:
🎉

jariz commented:
released @ 0.2.4 on npm
